tree/branch_node.go

```diff
@@ -0,0 +1,158 @@
+package tree
+
+import "fmt"
+
+// BranchNode represents a Node with an array of nodes and a sharedAddress
+// sharedAddress - part of the address that it links to
+// nodes are links to other nodes, which their addresses are suffixed by sharedAddress
+// [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, f ]
+//     [ x ] -> LeafNode
+//     [ not found ] -> EmptyNode
+//
+type BranchNode struct {
+	nodes         map[Unit]Node // TODO this can probably be an array if it help ?
+	sharedAddress []Unit
+	parent        Node
+	hash          []byte
+}
+
+// NewBranchNode returns a new BranchNode
+func NewBranchNode(sharedAddress []Unit, parent Node) Node {
+	return &BranchNode{
+		sharedAddress: sharedAddress,
+		nodes:         map[Unit]Node{},
+		parent:        parent,
+	}
+}
+
+// GetChild returns a child node
+// returns Node - one of it's children
+// if there isn't one
+// it returns EmptyNode - no node in this position
+//
+func (b *BranchNode) GetChild(key []Unit) Node {
+	// key - ABCDE
+	// b.sharedAddress - AC
+	// SharedPrefix() - A != b.sharedAddress
+	// if the node CAN'T exist in this prefix return an EmptyNode
+	if !EqualUnits(SharedPrefix(b.sharedAddress, key), b.sharedAddress) {
+		return NewEmptyNode(b, key)
+	}
+
+	// if the node CAN exist in this prefix but doesn't return an EmptyNode
+	node := b.nodes[FirstNonPrefix(b.sharedAddress, key)]
+	if node == nil {
+		return NewEmptyNode(b, key)
+	}
+
+	return node
+}
+
+// Insert adds a new node in the branch
+// if the node doesn't belong in the branch it inserts in the parent
+// if there's already a node on that position it creates a BranchNode and branches the position out
+// otherwise just inserts a LeafNode
+func (b *BranchNode) Insert(key []Unit, value []byte) {
+
+	// if the node CAN'T exist in this prefix request the parent to insert
+	if !EqualUnits(SharedPrefix(b.sharedAddress, key), b.sharedAddress) {
+		b.parent.Insert(key, value)
+		return
+	}
+
+	// if the node already exists then it's a new suffixed address
+	// needs a new branchNode
+	if node, ok := b.nodes[FirstNonPrefix(b.sharedAddress, key)]; ok {
+		newBranch := NewBranchNode(SharedPrefix(node.Key(), key), b)
+		newBranch.SetChild(node)
+		newBranch.Insert(key, value)
+
+		b.nodes[FirstNonPrefix(b.sharedAddress, key)] = newBranch
+		return
+	}
+
+	// all good, insert a LeafNode
+	b.nodes[FirstNonPrefix(b.sharedAddress, key)] = NewLeafNode(key, value, b)
+}
+
+// Delete
+// the delete request comes from a child node always (upwards direction)
+// if the child is a LeafNode, it requests the parent the deletion
+// if the child is a BranchNode,
+//     it either deletes + updates the tree
+//     or if there's no nodes left it deletes + requests parent the deletion
+//
+func (b *BranchNode) Delete(key []Unit) bool {
+	if _, ok := b.nodes[FirstNonPrefix(b.sharedAddress, key)]; ok {
```

Comment:
I think it'd be more clear to do:
```
if _, ok := b.nodes[FirstNonPrefix(b.sharedAddress, key)]; !ok {
    return false
}
...
```

Reply:
Ah, yeah I changed that earlier this week !  ☺️